File: backend/app/models/device.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, String, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Device(Base):
    __tablename__ = "devices"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    device_token: Mapped[str] = mapped_column(String(255), unique=True, nullable=False, index=True)
    owner_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)

    firmware_version: Mapped[str | None] = mapped_column(String(50), nullable=True)
    target_firmware_version: Mapped[str | None] = mapped_column(String(50), nullable=True)

    is_online: Mapped[bool] = mapped_column(Boolean, default=False)
    last_seen: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    ip_address: Mapped[str | None] = mapped_column(String(45), nullable=True)

    ui_descriptor: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    owner: Mapped["User"] = relationship("User", back_populates="devices")
    telemetry: Mapped[list["TelemetryRecord"]] = relationship("TelemetryRecord", back_populates="device", lazy="select", cascade="all, delete-orphan")

File: mobile/lib/models/device.dart
import 'ui_descriptor.dart';

class Device {
  final String id;
  final String name;
  final String deviceToken;
  final String? firmwareVersion;
  final bool isOnline;
  final String? lastSeen;
  final String? ipAddress;
  final UiDescriptor? uiDescriptor;

  Device({
    required this.id,
    required this.name,
    required this.deviceToken,
    this.firmwareVersion,
    required this.isOnline,
    this.lastSeen,
    this.ipAddress,
    this.uiDescriptor,
  });

  factory Device.fromJson(Map<String, dynamic> json) {
    return Device(
      id: json['id'] as String,
      name: json['name'] as String,
      deviceToken: json['device_token'] as String,
      firmwareVersion: json['firmware_version'] as String?,
      isOnline: json['is_online'] as bool? ?? false,
      lastSeen: json['last_seen'] as String?,
      ipAddress: json['ip_address'] as String?,
      uiDescriptor: json['ui_descriptor'] != null
          ? UiDescriptor.fromJson(json['ui_descriptor'] as Map<String, dynamic>)
          : null,
    );
  }
}
